File: C++/018. 4Sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        int size = nums.size();
        int left,right,val1,val2;
        sort(nums.begin(),nums.end());
        for(int i=0;i<size-3;i++)
        {
            for(int j=i+1;j<size-2;j++)
            {
                left=j+1;
                right=size-1;
                val1 = nums[i]+nums[j];
                while(left<right)
                {
                    val2 = nums[left]+nums[right];
                    if(val1+val2<target)
                    {
                        left++;
                    }
                    else if(val1+val2>target)
                    {
                        right--;
                    }
                    else
                    {
                        vector<int> sol(4,0);
                        sol[0] = nums[i];
                        sol[1] = nums[j];
                        sol[2] = nums[left];
                        sol[3] = nums[right];
                        res.push_back(sol);
                        while(left<right && nums[left]==sol[2]){left++;}
                        while(left<right && nums[right]==sol[3]){right--;}
                    }
                }
                while(j<size-2 && nums[j+1]==nums[j]){j++;}
                if(target-nums[i]-nums[j]<2*nums[j+1]){break;}
            }
            while(i<size-3 && nums[i+1]==nums[i]){i++;}
        }
        return res;
    }
};

File: python/018. 4Sum.py
Solution 1:
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        n = len(nums)
        if n == 0: return []
        nums.sort()
        res = []
        i = 0
        while i <= n-4:
            a = nums[i]
            if target - a < 3 * a: break
            j = i+1
            while j <= n-3:
                b = nums[j]
                e = target - a - b
                left, right = j + 1, n - 1
                if e < 2 * b: break
                while left < right:
                    if nums[left] + nums[right] == e:
                        res.append([a, b, nums[left], nums[right]])
                        left += 1
                        right -= 1
                        while left < right and nums[left] == nums[left-1]: left += 1
                        while left < right and nums[right] == nums[right+1]: right -= 1
                    elif nums[left] + nums[right] < e:
                        left += 1
                    else:
                        right -= 1
                while j <= n-3 and nums[j] == b: j += 1
            while i <= n-4 and nums[i] == a: i += 1
        return res

Solution 2:
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        n = len(nums)
        if n == 0: return []
        nums.sort()
        status = {}
        res = set()
        for i in range(n-1):
            for j in range(i+1, n):
                sum2 = nums[i] + nums[j]
                if sum2 not in status: status[sum2] = []
                status[sum2].append((i,j))
        for i in range(n-3):
            for j in range(i+1,n-2):
                t = target - nums[i] - nums[j]
                if t not in status: continue
                for x, y in status[t]:
                    if x > j: res.add((nums[i], nums[j], nums[x], nums[y]))
        res = [list(item) for item in res]
        return res